Tooltip fix for UI nodes (TypeLoadData)

aosyatnik opened the pull request:
### Purpose

Link to YouTrack:
[MAGN-5785](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-5785) DUI: Check correctness of Input/OutputParameters for NodeModelSearchElement

Without this task I can't finish Data Management tasks.
______________________

There are several types of nodes. One of them is node, that is inherited from `NodeModel`. We call them UI nodes, because they look different from others. E.g. 
![image](https://cloud.githubusercontent.com/assets/8158404/10016239/1dc73b1e-612e-11e5-8662-9c33275fc382.png)

None of them have right tooltip.
![image](https://cloud.githubusercontent.com/assets/8158404/10016255/3b6ea18e-612e-11e5-9b01-59f9af5d0c8e.png)![image](https://cloud.githubusercontent.com/assets/8158404/10016268/4ff35d20-612e-11e5-86e2-a8969681e99c.png)

Reason of problem is that in-coming and out-coming ports are generated actually after node is created. E.g. (this code is from Color Range)

``` c#
protected virtual void InitializePorts()
        {
            InPortData.Add(new PortData("colors", Resources.ColorRangePortDataColorsToolTip));
            InPortData.Add(new PortData("indices", Resources.ColorRangePortDataIndicesToolTip));
            InPortData.Add(new PortData("value", Resources.ColorRangePortDataValueToolTip));
            OutPortData.Add(new PortData("color", Resources.ColorRangePortDataResultToolTip));

            RegisterAllPorts();
        }
```

These nodes get their names and descriptions from `NodeNameAttribute` and `NodeDescriptionAttribute`. And I couldn't find better solution, then add 2 more attributes - `InputParametersAttribute`, `OutputParametersAttribute`.

Results:
![image](https://cloud.githubusercontent.com/assets/8158404/10016350/00776182-612f-11e5-89c5-4b6191d51a8f.png)
![image](https://cloud.githubusercontent.com/assets/8158404/10016358/10edb4a8-612f-11e5-8fd8-17a59a51ec73.png)

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@mjkkirschner 
@Benglin 

### FYIs

@Racel 
@jnealb

mjkkirschner commented on src/DynamoCore/Nodes/Attributes.cs:
```diff
@@ -208,4 +208,26 @@ public NodeObsoleteAttribute(string descriptionResourceID, Type resourceType)
     }
+
+    [AttributeUsage(AttributeTargets.Class)]
+    public class InputParametersAttribute : Attribute
```
what do you think about making these attributes tuple<string,type> that way we can display the type... just displaying the name is not that useful as the names are already present on the node, correct?

I'm also worried about the input parameters and input ports getting out of sync, potentially you could use get the name from the ports?

mjkkirschner commented on src/Libraries/CoreNodeModels/ColorRange.cs:
```diff
@@ -20,6 +20,8 @@ namespace DSCoreNodesUI
     [NodeCategory("Core.Color.Create")]
     [NodeDescription("ColorRangeDescription",typeof(DSCoreNodesUI.Properties.Resources))]
     [NodeSearchTags("ColorRangeSearchTags", typeof(DSCoreNodesUI.Properties.Resources))]
+    [InputParameters("colors", "indices", "value")]
```
Is there some way we can ensure that these names and the port names are in sync?

mjkkirschner commented:
@aosyatnik this looks good, but I think it would be more useful to include types as well if there are no problems with that. What do you think?

aosyatnik commented:
Hi, @mjkkirschner , I had a talk with @marimano and she showed how they work with UI nodes in Reach.
```c#
var newElement = nodeEl.CreateNode();
            
item.Parameters = newElement.InPorts.Select(elem => new LibraryItem.PortInfo
            {
                Name = elem.PortName,
                Type = elem.ToolTipContent.Split('.').Last(),
            });
```

They create instance of UI node and search for In and Out ports. So that they can be sure, that information about ports is always up-to-date. But I don't like one thing in this approach. 
```c#
Type = elem.ToolTipContent.Split('.').Last()`
```
That means if there is no dot in string, then we won't get Type of input port. Most of our tooltips don't have type at the end of the tooltip string. Or even more... We may have tooltips like "Useful tooltip. More text". In this case type will be considered as "More text".

So, what I meant to say. For now I see 2 options: create instance of node or try add InputParametersAttribute, OutputParametersAttribute.
1 option - good because we sure, that incoming data is up-to-date; bad because getting Type is thin spot.
2 option - good because there won't be any "strange" types; bad because it can be not synchronized with real node.

Let me know what you think. Thanks!

P.S. maybe we need to ask someone who knows Dynamo better...

pboyer commented:
@aosyatnik This is a tough one!  Unfortunately, it may also be quite a complex one to implement, affecting a lot of external code. @ikeough should know about this discussion.  

Because C# attributes are so limited (only allowing constant arguments), it's tough to provide them with any kind of structured data. I'd like to recommend this approach:

```csharp
[InPortNames("foo", "bar")]
[InPortDescriptions("super cool", "so nice!")]
[OutPortNames("baz", "bloo")]
[OutPortDescriptions("bee bop", "bla bla")]
public class Foo : NodeModel { ... }
```
This gives us the option of adding new fields (e.g. `InputTypes`). Unfortunately we have no way of confirming the attributes have the same number of parameters. The additional flexibility of being able to add more attributes later without modifying the additional ones is a big plus to me.

In the base class, `NodeModel`, we'll need something like this:

```csharp
public void RegisterInputPorts {
   // check if this method has already been called, if so, throw an exception

   // check the the InPortData field, if empty or length 0, check the attributes and use them to populate InPortData

   // create all input ports
}
```

We'd need to do something similar for OutPort's.

aosyatnik commented:
Hi, @pboyer , I understood your point of view. There is one thing, that may cause problems. If you want to move ports names, descriptions and types in attributes, then we can't use `Resource` file and as result there would be problems with localization of tooltips.
It's said, that `An attribute argument must be a constant expression, typeof expression or array creation expression of an attribute parameter type`, i.e. we can't write 
```c#
[InPortDescriptions(Resources.TooltipDescription)]
```
What do you think?

ikeough commented:
@aosyatnik There is a possible solution that gets us resource support. The names and descriptions attributes can be of a custom type, derived from Attribute, which looks up the resource. They do something like this here: http://stackoverflow.com/questions/1150874/c-sharp-attribute-text-from-resource-file.

I agree with @pboyer that backwards compatibility is required.

pboyer commented:
@ikeough Thanks for the pointer about attribute lookup. I believe that approach will work just fine.

@aosyatnik Can you go forward with the approach I described combined with @ikeough's addendum?

Here's what the `Attribute` class would look like:

```csharp 

public class InPortNamesAttribute: Attribute
{
    public InPortNamesAttribute(Type resourceType, params string[] resourceNames)
    {
    		PortNames= // look up resource names
    }

    public InPortNamesAttribute(params string[] portNames)
    {
    		PortNames= portNames;
    }

    public IEnumerable<string> PortNames{ get; private set; }
}
```

pboyer commented on src/DynamoCore/Nodes/Attributes.cs:
```diff
@@ -208,4 +208,173 @@ public NodeObsoleteAttribute(string descriptionResourceID, Type resourceType)
+    /// PortAttribute can process resource file (.resx) and load strings from this file.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class)]
+    public abstract class PortAttribute : Attribute
```
Let's call this `PortAttributeBase`.

pboyer replied:
Let's make this `internal`.

pboyer replied:
Instead of a base class, let's just make a small `internal` utility class that can read data from a resources type as done in the constructor. This gives us a bit more flexibility about the way this functionality evolves.

pboyer commented on src/DynamoCore/Nodes/Attributes.cs:
```diff
@@ -208,4 +208,173 @@ public NodeObsoleteAttribute(string descriptionResourceID, Type resourceType)
+        /// <param name="resourceNames">resource titles</param>
+        protected PortAttribute(Type resourceType, params string[] resourceNames)
+        {
+            if ((resourceType == null) || (resourceNames == null))
```
Throw `ArgumentNullExceptions`.

pboyer commented on src/DynamoCore/Nodes/Attributes.cs:
```diff
@@ -208,4 +208,173 @@ public NodeObsoleteAttribute(string descriptionResourceID, Type resourceType)
+    {
+        public InPortNamesAttribute(Type resourceType, params string[] resourceNames)
+            : base(resourceType, resourceNames)
+        { }
```
Here instead of calling the base class, we should call `PortAttributeLoader`.

pboyer commented on src/DynamoCore/Nodes/Attributes.cs:
```diff
@@ -208,4 +208,173 @@ public NodeObsoleteAttribute(string descriptionResourceID, Type resourceType)
+    [AttributeUsage(AttributeTargets.Class)]
+    public class InPortNamesAttribute : PortAttribute
+    {
+        public InPortNamesAttribute(Type resourceType, params string[] resourceNames)
```
Let's add documentation to each of these constructors.

pboyer commented on src/DynamoCore/Nodes/Attributes.cs:
```diff
@@ -208,4 +208,173 @@ public NodeObsoleteAttribute(string descriptionResourceID, Type resourceType)
+            : base(names)
+        { }
+
+        public IEnumerable<string> PortNames
```
Let's add documentation to this property.

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1129,6 +1146,20 @@ public void RegisterInputPorts()
+            // Used OutPortData.
+            if (OutPortData.Count > 0 && !isOutputPortsRegistered)
+            {
+                Warning(Properties.Resources.DeprecatedPortNamingStyleMessage, true);
```
Where is this warning emitted? I think we can leave this out for now.

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1092,6 +1094,20 @@ internal double GetPortVerticalOffset(PortModel portModel)
+            // Used InPortData.
+            if (InPortData.Count > 0 && !isInputPortsRegistered)
+            {
+                Warning(Properties.Resources.DeprecatedPortNamingStyleMessage, true);
```
Where is this warning emitted?

pboyer commented on test/Engine/FFITarget/DummyZeroTouchClass.cs:
```diff
@@ -16,4 +17,16 @@ public int FunctionWithoutDescription(int a)
+    [OutPortNames("out1", "out2", "out3")]
+    [OutPortTypes("1", "2", "3")]
+    [OutPortDescriptions("desc")]
+    public class DummyZeroTouchClass1
```
It's erroneous to put this on a zero touch class - these attributes should only be applied to `NodeModel` sub classes. We should put this on a dummy class called `DummyNodeModel`. These port descriptions should generally not be applied to

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1129,6 +1146,20 @@ public void RegisterInputPorts()
+            // Used port Attributes.
+            if (!isOutputPortsRegistered)
+            {
+                OutPortData.AddRange(GetPortsDataFromAttributes(PortType.Output));
```
Can we use a local variable instead of mutating `OutPortData`? We should probably apply the `[Obsolete]` attribute to both the `InPortData` property and the `OutPortData` property.

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1092,6 +1094,20 @@ internal double GetPortVerticalOffset(PortModel portModel)
         /// </summary>
         public void RegisterInputPorts()
```
Let's also make this obsolete. 
```
[Obsolete("RegisterInputPorts is deprecated, please use the InPortNamesAttribute, InPortDescriptionsAttribute, and InPortTypesAttribute instead.")]
```

aosyatnik replied:
We still use `RegisterInputPorts`, it is not `Obsolete` in fact. We just should make it private/internal. But we can't, because `RegisterInputPorts` is used in many other places.

pboyer replied:
We mark code obsolete to indicate that we will eventually remove it. Not that we're necessarily removing it right now.

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -42,6 +42,8 @@ public abstract class NodeModel : ModelBase, IRenderPackageSource<NodeModel>, ID
         private string description;
         private string persistentWarning = "";
+        private bool isInputPortsRegistered;
+        private bool isOutputPortsRegistered;
```
`areOutputPortsRegistered`, `areInputPortsRegistered`.

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1159,6 +1190,63 @@ public void RegisterOutputPorts()
+            }
+
+            if (names == null)
+                return new List<PortData>();
```
Please surround with curly braces as suggested in the style guide.

pboyer commented on test/DynamoCoreTests/TestZeroTouchClass.cs:
```diff
@@ -72,5 +74,107 @@ public void DescriptionTest()
             node = new DSFunction(fucDescriptor);
             Assert.AreEqual(fucDescriptor.Signature, node.Description);
         }
+
```
These tests should not be in `TestZeroTouchClass`. These attributes should not be applied to zero touch classes. They tests should be among other `NodeModel` tests.

pboyer commented:
@aosyatnik I have a few comments here. Sorry, I have to be a bit more stringent here. There are many external consumers of the `NodeModel` class and we have to make sure the code is clear about where things are headed.

aosyatnik commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1092,10 +1096,27 @@ internal double GetPortVerticalOffset(PortModel portModel)
+            if (InPortData.Count > 0)
+            {
+                inputs.AddRange(InPortData);
+                Log(String.Concat(Name, ": ", Properties.Resources.DeprecatedPortNamingStyleMessage));
```
Is it what you wanted to use?

pboyer replied:
Let's remove this log. The developer should get compile time warnings, which is sufficient.

pboyer commented on src/DynamoCore/Properties/Resources.en-US.resx:
```diff
@@ -624,4 +624,22 @@ value : bool = false</value>
   </data>
+  <data name="DeprecatedPortNamingStyleMessage" xml:space="preserve">
+    <value>You used old style of setting node ports. Please, use new style based on adding port attributes.</value>
+  </data>
```
As we're removing the `Log` call, we can also remove this.

pboyer commented on src/DynamoCore/Utilities/ResourceLoader.cs:
```diff
@@ -0,0 +1,69 @@
+                    throw new InvalidOperationException(Resources.ResourcePropertyIsNotStringTypeMessage);
+                }
+                titles.Add((string)property.GetValue(null, null));
+            }
```
This should just be calling the first overload, not duplicating the code entirely...

pboyer commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1159,6 +1198,68 @@ public void RegisterOutputPorts()
+        /// Tries to load ports names and descriptions from attributes.
+        /// </summary>
+        /// <param name="portType">Input or Output port type</param>
+        private IEnumerable<PortData> GetPortsDataFromAttributes(PortType portType)
```
`GetPortDataFromAttributes`

pboyer commented:
@aosyatnik This is very close. I'd also like you to do a run on the self-service CI to ensure no regressions. We'll also need to update the documentation to reflect these changes: @ikeough Do you know what we'll need to update?

ksobon commented:
This is an interesting development. Please note that this fix would address a pretty wide range of nodes that includes all of the nodes derived from <code>DropDown</code> - or at least I hope as I have not been able to add proper descriptions for input and output so far. 

![untitled-2](https://cloud.githubusercontent.com/assets/529781/10413961/527247ba-6f8d-11e5-8d82-0fc651f274a1.jpg)


![untitled-1](https://cloud.githubusercontent.com/assets/529781/10413956/1649be58-6f8d-11e5-8a8c-78642910faa9.jpg)

Will these fixes be implemented for next stable?

pboyer commented:
@ksobon This will likely not be in the next stable as it will require refactoring a large number of nodes to take advantage of the new attribute based input/output creation description mechanism
